### PR TITLE
feat(cb2-4228): amend custom defects

### PR DESCRIPTION
**Pull request by Ruairidh-BJSS**
## Amend Custom Defects (Specialist Tests)

_This story covers functionality that will allow the user to amend custom defect fields for a test record and special test types that don’t use defects taxonomy_
[link to ticket number](https://dvsa.atlassian.net/browse/CB2-4228)

## Checklist

- [ ] Branch is rebased against the latest develop/common
- [ ] Necessary `id` required prepended with `"test-"` have been checked with automation testers and added
- [ ] Code and UI has been tested manually after the additional changes
- [ ] PR title includes the JIRA ticket number
- [ ] Squashed commits contain the JIRA ticket number
- [ ] Link to the PR added to the repo
- [ ] Delete branch after merge

### CODE REVIEW

**Review by mcadenas-bjss — DISMISSED** [review]

File: src/app/forms/custom-sections/custom-defect/custom-defect.component.html [hunk 1]
```diff
@@ -0,0 +1,42 @@
+<ng-template [ngTemplateOutlet]="isEditing ? editCustomDefect : viewCustomDefect" [ngTemplateOutletContext]="{ data: form }"></ng-template>
```
Comment (tahayazici) [inline src/app/forms/custom-sections/custom-defect/custom-defect.component.html:1]: Very nice, very clean template. Nicely put together 👌

File: src/app/forms/custom-sections/custom-defect/custom-defect.component.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
+import { CustomFormGroup } from '@forms/services/dynamic-form.types';
+
+@Component({
+  selector: 'app-custom-defect[index]',
+  templateUrl: './custom-defect.component.html',
+  styleUrls: ['./custom-defect.component.scss']
+})
+export class CustomDefectComponent implements OnInit {
+  @Input() form!: CustomFormGroup;
```
Comment (tahayazici) [inline src/app/forms/custom-sections/custom-defect/custom-defect.component.ts:10]: If you always expect the `form`, can you please include it in the selector?

File: src/app/forms/custom-sections/custom-defect/custom-defect.component.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  isEditingCustomDefect: boolean = false;
+  sectionFormValues!: CustomFormGroup;
+
+  ngOnInit(): void {
+    this.updateSectionFormValues();
+  }
+
+  updateSectionFormValues(): void {
+    this.sectionFormValues = this.form.value;
+  }
```
Comment (tahayazici) [inline src/app/forms/custom-sections/custom-defect/custom-defect.component.ts:24]: It's just a one-liner, no need to be put on a separate method, unless the template references it.
```suggestion
  ngOnInit(): void {
    this.sectionFormValues = this.form.value;
  }
```

Comment (gjulien-bjss) [inline src/app/forms/custom-sections/custom-defect/custom-defect.component.ts:24]: Is there a point to making this a separate method if it's only being used in the `ngOnInit` lifecycle hook?
```suggestion
  ngOnInit(): void {
    this.sectionFormValues = this.form.value;
  }
```

File: src/app/forms/custom-sections/custom-defects/custom-defects.component.html [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      <app-custom-defect
+        *ngFor="let customDefect of customDefects; let i = index; trackBy: trackByFn"
+        [isEditing]="isEditing"
+        [form]="getCustomDefectForm(i)"
+        [index]="i"
+        (removeCustomDefect)="handleRemoveDefect($event)"
+      ></app-custom-defect>
+    </ng-container>
+  </ng-container>
+</ng-container>
```
Comment (tahayazici) [inline src/app/forms/custom-sections/custom-defects/custom-defects.component.html:15]: You can remove the nesting here
```suggestion
<ng-container *ngIf="form && defectCount; else noDefects" [formGroup]="form">
  <h2 class="govuk-heading-m">Custom Defects</h2>

  <app-custom-defect
    *ngFor="let customDefect of customDefects; let i = index; trackBy: trackByFn"
    [isEditing]="isEditing"
    [form]="getCustomDefectForm(i)"
    [index]="i"
    (removeCustomDefect)="handleRemoveDefect($event)"
  ></app-custom-defect>
</ng-container>
```

File: src/app/forms/custom-sections/custom-defect/custom-defect.component.ts (hunk 3 above)
Comment (gjulien-bjss) [reply src/app/forms/custom-sections/custom-defect/custom-defect.component.ts:24]: Also, where is `sectionFormValues` being used?

File: src/app/forms/custom-sections/custom-defects/custom-defects.component.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    this.formSubscription = this.form.cleanValueChanges.subscribe(event => {
+      this.formChange.emit(event);
+    });
+  }
+
+  ngOnDestroy(): void {
+    this.formSubscription.unsubscribe();
+  }
+
+  get customDefectsForm() {
```
Comment (tahayazici) [inline src/app/forms/custom-sections/custom-defects/custom-defects.component.ts:35]: Very neat component, good stuff 👍

File: src/app/forms/custom-sections/custom-defects/custom-defects.component.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  get defectCount() {
+    return this.customDefectsForm?.controls.length;
+  }
+
+  get customDefects(): CustomDefects {
+    return this.customDefectsForm.controls.map(control => {
+      const formGroup = control as CustomFormGroup;
+      return formGroup.getCleanValue(formGroup) as CustomDefect;
+    });
+  }
```
Comment (gjulien-bjss) [inline src/app/forms/custom-sections/custom-defects/custom-defects.component.ts:56]:
```suggestion
  get customDefects(): CustomDefects {
    return this.customDefectsForm.controls.map(control => 
     (formGroup as CustomFormGroup).getCleanValue(formGroup as CustomFormGroup) as CustomDefect;
    );
  }
```

**Review by kpearmanbjss — APPROVED** [review]

**Review by leifkemp-bjss — APPROVED** [review]

**Pull request merged**